File: Dec2020/9Dec/AverageOfLevelsInBinaryTree.cpp
/*
Problem name: 637. Average of Levels in Binary Tree
Problem link: https://leetcode.com/problems/average-of-levels-in-binary-tree/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int size=q.size();
            double sum=0;
            for(int i=0;i<size;++i){
                TreeNode* node = q.front();
                sum += node->val;
                if(node->left)
                    q.push(node->left);
                if(node->right)
                    q.push(node->right);
                q.pop();
            }
            ans.push_back(sum/size);
        }
        return ans;
    }
};

File: Dec2020/9Dec/AverageOfLevelsInBinaryTree.java
/*
Problem Name: 637. Average of Levels in Binary Tree
Problem Link: https://leetcode.com/problems/average-of-levels-in-binary-tree/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> ans = new ArrayList<Double>();
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        q.add(root);
        q.add(null);
        double sum = 0, cnt = 0;
        while(!q.isEmpty()){
            TreeNode t = q.remove();
            if(t == null){
                if(q.peek() == null)
                    break;
                else{
                    ans.add(sum/cnt);
                    sum = 0;
                    cnt = 0;
                    q.add(null);
                }
            }
            else{
                sum += t.val;
                cnt++;
                if(t.left != null)
                    q.add(t.left);
                if(t.right != null)
                    q.add(t.right);
            }
        }
        ans.add(sum/cnt);
        return ans;
    }
}
